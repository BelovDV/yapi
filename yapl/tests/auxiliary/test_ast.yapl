.fn square(a: .int) -> .int
  * a a

.fn rif(ar: .int)
  = a + ar 2
  = b + * a 2 - ar 2

. .fn read_and_sum_squares(amount: .int) -> .int
.   { gen amount }.map(square).fold(+)
.     .read(.int)

. .fn fib(arg: .unsigned) -> .unsigned
.   { req .unsigned }[arg]
.         0 => 0
.         1 => 1
.     + [-1] [-2]

. .fn read_and_sum(arg: .unsigned) -> .int
.   { gen arg }.fold(+)
.     .read(.int)

. .fn read_and_check_if_all_more_3_mod_7(arg: .unsigned) -> .string
.   .fn mod_7(a: .int) -> .int
.     % a 7
.   .fn more_3(a: .int) -> .bool
.     > a 3
.   = result { gen arg }.map(more_3).fold(+)
.     mod_2 .read(.int)
.   match result
.     .true
.       "yes"
.     .false
.       "no"
.   result


. .fn process(type: .type, f: .fn) -> ()
.   = a .read(type)
.   = b f(a)
.   .print(b)

. .fn main(args: .placeholder)
.   process(.unsigned, fib)
.   process(.unsigned, read_and_sum)
.   process(.unsigned, read_and_check_if_all_more_3_mod_7)
