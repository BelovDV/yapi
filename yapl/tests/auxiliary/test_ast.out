[
    Function {
        name: Symbol(1: "square"),
        args: [
            Symbol(2: "a"),
        ],
        stmts: [
            Mapped {
                func: Symbol(6: "*"),
                args: [
                    Symbol(
                        Symbol(2: "a"),
                    ),
                    Symbol(
                        Symbol(2: "a"),
                    ),
                ],
            },
        ],
    },
    Function {
        name: Symbol(7: "rif"),
        args: [
            Symbol(8: "ar"),
        ],
        stmts: [
            Mapped {
                func: Symbol(9: "="),
                args: [
                    Symbol(
                        Symbol(2: "a"),
                    ),
                    Mapped {
                        func: Symbol(10: "+"),
                        args: [
                            Symbol(
                                Symbol(8: "ar"),
                            ),
                            ConstantI(
                                2,
                            ),
                        ],
                    },
                ],
            },
            Mapped {
                func: Symbol(9: "="),
                args: [
                    Symbol(
                        Symbol(11: "b"),
                    ),
                    Mapped {
                        func: Symbol(10: "+"),
                        args: [
                            Mapped {
                                func: Symbol(6: "*"),
                                args: [
                                    Symbol(
                                        Symbol(2: "a"),
                                    ),
                                    ConstantI(
                                        2,
                                    ),
                                ],
                            },
                            Mapped {
                                func: Symbol(12: "-"),
                                args: [
                                    Symbol(
                                        Symbol(8: "ar"),
                                    ),
                                    ConstantI(
                                        2,
                                    ),
                                ],
                            },
                        ],
                    },
                ],
            },
        ],
    },
]